### PR TITLE
[SYCL] Prework before protecting access to the event objects with a mutex

### PR DESCRIPTION
**Pull request by againull**
* Currently resetCommandList() calls cleanup() and vice versa which makes it difficult to guard access to pi_objects processed by these functions leading to locks of mutexes in different order (causing dead lock). So, don't call resetCommandList from event->cleanup(), we can call resetCommandLists() from piEventsWait() directly instead. Also, remove VM_BIND workaround which is incorrect because currently we perform batching of the copy command lists. 
* Remove pointer to command list (ZeCommandList)  from pi_event. Command list references event as well and this introduces difficulty: we need to lock a queue and then an event to be able to process events in the command list. And vice versa to lock an event and then a queue to update ZeCommandList when cleaning up an event.
* Perform cleanup and release of events outside of queue lock. Callers of cleanup and EventRelease must no lock any mutexes.
As a related change remove LockedQueue parameter from cleanup functions. Currently LockedQueue parameter is used several functions for cleaning resources. It specifies which queue is locked in the caller of these functions to not lock it twice. Problem is that some of the functions internally process events in a loop which can be from different queues and these queues are locked as part of this processing. It leads to the problem that we can have nested locks of different queues and this is a classic deadlock situation.
* Make cleanup() and resetCommandLists() functions free. The resetCommandLists() function performs release of events in command lists and we release the queue if event's refcount reaches 0, so it is possible that queue can be destructed inside this function. That's why it should be free function and not a method of the queue. Same for cleanup().

### CODE REVIEW

**Comment by againull** [comment]
Sergey, I've made changes per our discussion as a prework to protect access to events with a mutex.
Since they logically try to solve related problems and don't touch a lot of code, I've put them in one PR.

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 1]
```diff
@@ -1454,8 +1456,10 @@ pi_result _pi_queue::executeCommandList(pi_command_list_ptr_t CommandList,
       // command-list itself. This host-visible event will not be
       // waited/released by SYCL RT, so it must be destroyed after all events in
       // the batch are gone.
-      PI_CALL(piEventRelease(HostVisibleEvent));
-      PI_CALL(piEventRelease(HostVisibleEvent));
+      // We know here that refcount is more than 2 so decrement refcount
+      // manually to avoid calling piEventRelease under queue lock.
+      assert(HostVisibleEvent->RefCount > 2);
+      HostVisibleEvent->RefCount -= 2;
```
Comment (aelovikov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:1462]: That would make traces much harder to read as the decrement wouldn't be shown there.

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 2]
```diff
@@ -30,10 +30,11 @@
 
 extern "C" {
 // Forward declarartions.
-static pi_result EventRelease(pi_event Event, pi_queue LockedQueue);
-static pi_result QueueRelease(pi_queue Queue, pi_queue LockedQueue);
+static pi_result EventRelease(pi_event Event);
+static pi_result QueueRelease(pi_queue Queue);
 static pi_result EventCreate(pi_context Context, pi_queue Queue,
                              bool HostVisible, pi_event *RetEvent);
+static pi_result cleanup(pi_event Event);
```
[the remark this answers is not in the document]
Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: Good call, fixed.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 1 above)
Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: I agree, fixed.

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 3]
```diff
@@ -890,21 +887,16 @@ pi_result _pi_queue::resetCommandList(pi_command_list_ptr_t CommandList,
...
   }
 
-  // Finally release/cleanup all the events in this command list.
-  // Note, we don't need to synchronize the events since the fence
-  // synchronized above already does that.
   auto &EventList = CommandList->second.EventList;
+  // We don't need to synchronize the events since the fence
+  // synchronized above already does that.
   for (auto &Event : EventList) {
     Event->Completed = true;
```
[the remark this answers is not in the document]
Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: Indeed, this can be moved, fixed.

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 4]
```diff
@@ -5606,19 +5602,18 @@ pi_result piEventRetain(pi_event Event) {
   return PI_SUCCESS;
 }
 
-pi_result piEventRelease(pi_event Event) {
-  return EventRelease(Event, nullptr);
-}
+pi_result piEventRelease(pi_event Event) { return EventRelease(Event); }
```
[the remark this answers is not in the document]
Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: I agree, fixed.

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 5]
```diff
@@ -1126,6 +1119,100 @@ _pi_queue::_pi_queue(std::vector<ze_command_queue_handle_t> &ComputeQueues,
   CopyCommandBatch.QueueBatchSize = ZeCommandListBatchCopyConfig.startSize();
 }
 
+// Perform any necessary cleanup after an event has been signalled.
+// This currently makes sure to release any kernel that may have been used by
+// the event, updates the last command event in the queue and cleanes up all dep
+// events of of the event.
```
Comment (aelovikov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:1125]: typo "of of"

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 6]
```diff
@@ -1454,6 +1547,9 @@ pi_result _pi_queue::executeCommandList(pi_command_list_ptr_t CommandList,
       // command-list itself. This host-visible event will not be
       // waited/released by SYCL RT, so it must be destroyed after all events in
       // the batch are gone.
+      // We know that refcount is more than 2 so calling piEventRelease is safe
+      // because it locks the queue only if refcount of the event is zero.
+      assert(HostVisibleEvent->RefCount > 2);
```
Comment (aelovikov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:1552]: Why can't it be exactly 2?

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 7]
```diff
@@ -1134,28 +1119,136 @@ _pi_queue::_pi_queue(std::vector<ze_command_queue_handle_t> &ComputeQueues,
...
+    // Lock automatically releases when this goes out of scope.
+    std::scoped_lock Lock(AssociatedQueue->Mutex);
+
+    // If this event was the LastCommandEvent in the queue, being used
+    // to make sure that commands were executed in-order, remove this.
+    // If we don't do this, the event can get released and freed leaving
+    // a dangling pointer to this event.  It could also cause unneeded
+    // already finished events to show up in the wait list.
+    if (AssociatedQueue->LastCommandEvent == Event) {
+      AssociatedQueue->LastCommandEvent = nullptr;
```
Comment (aelovikov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:1170]: If possible, it might be better to wrap `LastCommandEvent` into `std::atomic` and do `compare_exchange_strong` without mutex lock.

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 8]
```diff
@@ -1134,28 +1119,136 @@ _pi_queue::_pi_queue(std::vector<ze_command_queue_handle_t> &ComputeQueues,
...
+    }
+  }
+
+  // Release this event since we explicitly retained it on creation and
+  // association with queue. Events which don't have associated queue doesn't
+  // require this release because it means that they are not created using
+  // createEventAndAssociateQueue, i.e. additional retain was not made.
+  if (AssociatedQueue)
+    PI_CALL(piEventRelease(Event));
```
Comment (aelovikov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:1179]: Can we move it between lines 1171/1172 that are already guarded by the same condition?

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 9]
```diff
@@ -1134,28 +1119,136 @@ _pi_queue::_pi_queue(std::vector<ze_command_queue_handle_t> &ComputeQueues,
...
       }
     }
   }
+  for (auto Event : EventListToCleanup) {
+    // We don't need to synchronize the events since the fence
+    // synchronized above already does that.
+    Event->Completed = true;
+
+    PI_CALL(cleanup(Event));
+    // These events were just removed from command list, so dercement ref count
```
Comment (aelovikov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:1248]:
```suggestion
    // These events were just removed from command list, so decrement ref count
```

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 6 above)
Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1554]: It can't be exactly 2 because I check that EventList of the command list is not empty above, i.e. after createEventAndAssociateQueue ref count is 2 and then +1 for each event in the EventList. This code makes sense only if EventList is not empty, more over before this change it was possible to get a situation when deleted Event is dereferenced if EventList is empty.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 5 above)
Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: Fixed.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 8 above)
Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1179]: Moved.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 9 above)
Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: Fixed.

**Review by aelovikov-intel — COMMENTED** [review]
I don't have any more comments. @smaslov-intel , I'd expect Arthur is awaiting your review here.

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 10]
```diff
@@ -1134,28 +1119,137 @@ _pi_queue::_pi_queue(std::vector<ze_command_queue_handle_t> &ComputeQueues,
   CopyCommandBatch.QueueBatchSize = ZeCommandListBatchCopyConfig.startSize();
 }
 
+// Perform any necessary cleanup after an event has been signalled.
+// This currently makes sure to release any kernel that may have been used by
+// the event, updates the last command event in the queue and cleanes up all dep
+// events of the event.
+// The caller must not lock any mutexes.
+static pi_result cleanup(pi_event Event) {
```
[the remark this answers is not in the document]
Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: Ok, done.

File: sycl/plugins/level_zero/pi_level_zero.hpp [hunk 11]
```diff
@@ -891,11 +891,7 @@ struct _pi_queue : _pi_object {
   // needs to be set to true. The caller must verify that this command list and
   // fence have been signalled.
   pi_result resetCommandList(pi_command_list_ptr_t CommandList,
-                             bool MakeAvailable);
-
-  // Reset signalled command lists in the queue and put them to the cache of
-  // command lists. A caller must not lock the queue mutex.
-  pi_result resetCommandLists();
+                             bool MakeAvailable, std::vector<_pi_event *> &);
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.hpp:894]: Document the new argument, and update function comment.

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 12]
```diff
@@ -1134,28 +1119,44 @@ _pi_queue::_pi_queue(std::vector<ze_command_queue_handle_t> &ComputeQueues,
   CopyCommandBatch.QueueBatchSize = ZeCommandListBatchCopyConfig.startSize();
 }
 
+static pi_result cleanup(pi_event);
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:1122]: Why did you change this to be a free function instead of being a member of `_pi_event`?

Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:1122]: Why did you change this to be a free function instead of being a member of `_pi_event`?

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1122]: I now see your comment in the description:

> Make cleanup() and resetCommandLists() functions free. The resetCommandLists() function performs release of events in command lists and we release the queue if event's refcount reaches 0, so it is possible that queue can be destructed inside this function. That's why it should be free function and not a method of the queue. Same for cleanup().

What's wrong with `cleanup()` being a PI event's member function? How is this solved with it being a free function?

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 13]
```diff
@@ -1134,28 +1119,44 @@ _pi_queue::_pi_queue(std::vector<ze_command_queue_handle_t> &ComputeQueues,
...
   }
+  for (auto Event : EventListToCleanup) {
+    // We don't need to synchronize the events since the fence
+    // synchronized above already does that.
+    Event->Completed = true;
+
+    PI_CALL(cleanup(Event));
+    // These events were just removed from command list, so decrement ref count
+    // (it was incremented when they were added to the command list).
+    PI_CALL(piEventRelease(Event));
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:1158]: Why did you move this out of cleanup()?

Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:1158]: Why did you move this out of cleanup()?

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 14]
```diff
@@ -1431,7 +1432,8 @@ pi_result _pi_queue::executeCommandList(pi_command_list_ptr_t CommandList,
     // event that anybody waiting for any event in the batch will
     // really be using.
     //
-    if (EventsScope == LastCommandInBatchHostVisible) {
+    if (EventsScope == LastCommandInBatchHostVisible &&
+        !CommandList->second.EventList.empty()) {
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:1436]: Why is this extra check? Just an optimization?

Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:1436]: Why is this extra check? Just an optimization?

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1436]: I see explanation in conversation just below, and it makes sense but please add a source comment too.

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1436]: I see explanation in conversation just below, and it makes sense but please add a source comment too.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 12 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1122]: Isn't it the same thing? Essentially an "event" should be considered dead after piEventRelease.
That's actually true for every "release" interface that it deletes the object if last reference to it is released.
So this release should be the last thing you do with an object.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 13 above)
Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1158]: This piEventRelease is moved here from resetCommandList, not from cleanup. We reset command list above  and removed these events from its EventList container. So we can decrement refcount.
The idea was to move cleanup and piEventRelease functions out of queue lock because they can possibly call piQueueRelease which locks the queue. 
I can move piEventRelease back to the resetCommandList if https://github.com/intel/llvm/pull/6192 gets merged.
So it is probably better to merge https://github.com/intel/llvm/pull/6192  first, then I can update this patch.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 12 above)
Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1122]: Just to clarify, this is not so important change and it is a matter of taste. It is highly recommended to avoid "delete this" that's why I've made this change because I was trying to improve things.  But it is ok to do "delete this" as long as we are careful: https://isocpp.org/wiki/faq/freestore-mgmt#delete-this

If you insist on having this as the method of pi_event I can revert this change.

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 15]
```diff
@@ -5413,127 +5430,105 @@ pi_result piEventGetProfilingInfo(pi_event Event, pi_profiling_info ParamName,
...
-      if (Queue->RefCount > 0) {
-        auto it = Queue->CommandListMap.find(ZeCommandList);
-        if (it == Queue->CommandListMap.end()) {
-          die("Missing command-list completition fence");
-        }
+// This currently makes sure to release any kernel that may have been used by
+// the event, updates the last command event in the queue and cleanes up all dep
+// events of the event.
+// The caller must not lock any mutexes.
+static pi_result cleanup(pi_event Event) {
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:5440]: if you prefer to keep this as a free function please rename it:
```suggestion
static pi_result CleanupCompletedEvent(pi_event Event) {
```
```suggestion
static pi_result cleanup(pi_event Event) {
```

Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:5440]: if you prefer to keep this as a free function please rename it:
```suggestion
static pi_result CleanupCompletedEvent(pi_event Event) {
```
```suggestion
static pi_result cleanup(pi_event Event) {
```

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 6 above)
Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1461]: It's not just an optimization. Problem is that if even list is empty then HostVisible event will be deleted after the second call to piEventRelease and we use it afterwards https://github.com/intel/llvm/pull/6179/files#diff-15dd1eb076d2164bd9e87d9057f05f652a716498e8cdf5975e564c65309a0985R1465-R1473.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 12 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1122]: No, I am fine with free-standing. but please rename it then

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 6 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1461]: I see. Please add a comment about this where you check for non-empty list

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 16]
```diff
@@ -5607,18 +5612,14 @@ pi_result piEventRetain(pi_event Event) {
...
-static pi_result EventRelease(pi_event Event, pi_queue LockedQueue) {
   PI_ASSERT(Event, PI_INVALID_EVENT);
   if (!Event->RefCount) {
     die("piEventRelease: called on a destroyed event");
   }
 
   if (--(Event->RefCount) == 0) {
     if (!Event->CleanedUp)
-      Event->cleanup(LockedQueue);
+      cleanup(Event);
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:5622]: It looks like this call to "cleanup" here is wrong, or at lease redundant.
Since refcount came to 0 already, it means the "cleanup" was already done?
What do you think?

Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:5622]: It looks like this call to "cleanup" here is wrong, or at lease redundant.
Since refcount came to 0 already, it means the "cleanup" was already done?
What do you think?

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 17]
```diff
@@ -6046,8 +6047,7 @@ pi_result _pi_queue::synchronize() {
             (ImmCmdList->first, zeEvent, 0, nullptr));
     ZE_CALL(zeHostSynchronize, (zeEvent));
     Event->Completed = true;
-    Event->ZeCommandList = nullptr;
-    PI_CALL(Event->cleanup(Queue));
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:6050]: when will this event be cleaned up then?

Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:6050]: when will this event be cleaned up then?

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 13 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1158]: I see, thanks. I think it is clearer if we release goes in the same place where the reference is removed.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 16 above)
Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.cpp:5622]: I completely agree with that. I was trying to make this change here some time ago: https://github.com/intel/llvm/pull/6073

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 17 above)
Comment (againull) [reply sycl/plugins/level_zero/pi_level_zero.cpp:6050]: This event doesn't need a cleanup. It is created using EventCreate and not using CreateEventAndAssociateQueue, i.e. it doesn't have associated queue, kernel or dep events. It is created and released inside of this scope and doesn't go outside, so no cleanup is needed and it can be just released.
